java-checks/src/main/java/org/sonar/java/checks/WrongAssignmentOperatorCheck.java

```diff
@@ -47,12 +47,17 @@ public void visitNode(Tree tree) {
     SyntaxToken expressionFirstToken = aeTree.expression().firstToken();
     SyntaxToken variableLastToken = aeTree.variable().lastToken();
     if (isSuspiciousToken(expressionFirstToken)
+      && !isSingleNegationAssignment(expressionFirstToken, aeTree)
       && noSpacingBetween(operatorToken, expressionFirstToken)
       && !noSpacingBetween(variableLastToken, operatorToken)) {
       reportIssue(operatorToken, expressionFirstToken, "Was \"" + expressionFirstToken.text() + "=\" meant instead?");
     }
   }
 
+  private static boolean isSingleNegationAssignment(SyntaxToken firstToken, AssignmentExpressionTree aeTree) {
+    return "!".equals(firstToken.text()) && (aeTree.parent() == null || !aeTree.parent().is(Tree.Kind.ASSIGNMENT));
+  }
+
   private static boolean noSpacingBetween(SyntaxToken firstToken, SyntaxToken secondToken) {
     return firstToken.column() + firstToken.text().length() == secondToken.column();
```

Comment:
I just realized that this could generate FPs when placed on different lines:
```java
target =
        +num;
```
We should check that both token are sharing the same line.

Reply:
oh nice. good catch. I'll update accordingly